### ISSUE TITLE
Zero encountered in variance in NormalizationLayer leads to backend.epsilon() being used which is a questionable choice.

### ISSUE THREAD

**Issue by yCobanoglu** [keras-team/keras#20011]
Hi,

as mentioned in the title. If the variance contains a zero entry than the standardization uses backend.epsilon() as variance. Dividing the sample by a very small value leads everything to blow up essentially. You can check for example `class sklearn.preprocessing.StandardScaler` implementation where if variance is zero it is set to one (or something along these lines). I encountered this problem on mnist where the variance of the test set ends up as 14693871000.0 after normalization.

```python
import keras
import tensorflow as tf
import tensorflow_datasets as tfds

train_ds = tfds.load("mnist", split="train", shuffle_files=True)
test_ds = tfds.load("mnist", split="test")

_X_train = next(iter(train_ds.batch(len(train_ds))))["image"]
_X_train = tf.reshape(_X_train, [_X_train.shape[0], -1])

_X_test = next(iter(test_ds.batch(len(train_ds))))["image"]
_X_test = tf.reshape(_X_test, [_X_test.shape[0], -1])

standardize = keras.layers.Normalization(axis=-1)
standardize.adapt(_X_train)
_X_train = standardize(_X_train)
_X_test = standardize(_X_test)

test_var = tf.math.reduce_variance(_X_test)
print(standardize.variance) #14693871000.0
```

```python
from keras.src.datasets import mnist
from sklearn.preprocessing import StandardScaler

(X_train, y_train), (X_test, y_test) = mnist.load_data()

X_train = X_train.reshape(60000, 784)
X_test = X_test.reshape(10000, 784)

X_train = X_train.astype("float32")
X_test = X_test.astype("float32")

standardize = StandardScaler().fit(X_train)

X_train = standardize.transform(X_train)
X_test = standardize.transform(X_test)

print(np.var(X_test)) # 0.9161259

```
Also is there any plans to support something like `class sklearn.preprocessing.StandardScaler.partial_fit()`. This would be good so one doesn't have to compute the mean and var for the complete dataset but can incrementally update it during iterations through the dataset.

**Comment by sachinprasadhs** [comment]
@grasskin , Here is the issue repro gist 
[![](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.sandbox.google.com/gist/sachinprasadhs/1a0b6c16c2a80f6e91cbbb6aa9099e9e/20011.ipynb)

**Comment by grasskin** [comment]
Thank you @sachinprasadhs! Interesting, digging more into sklearn's implementation they check whether variance is 0 and set standard deviation to 1 https://github.com/scikit-learn/scikit-learn/blob/d20e0b9abc4a4798d1fd839db50b19c01723094e/sklearn/preprocessing/_data.py#L82 as called in https://github.com/scikit-learn/scikit-learn/blob/d20e0b9abc4a4798d1fd839db50b19c01723094e/sklearn/preprocessing/_data.py#L1013

Looking at our Keras 2 implementation Keras has long opted for just avoiding dividing by 0 https://github.com/keras-team/tf-keras/blob/8eb5627387f83c2ebd31b31aad92f42ad0a8dcdb/tf_keras/layers/preprocessing/normalization.py#L360.

Backwards compatibility on this is complicated, a possible middle ground is to add a `ignore_zeros` parameter.
@yCobanoglu would you expect your data to be mean normalized (`x-mean`) but not scaled (divided by `sqrt(var)`) in these cases?

On `partial_fit` a Keras equivalent `partial_adapt` might be interesting - we would have to track how many batches we've seen and take into account previous computation. Is this a feature you would have found helpful?

**Comment by yCobanoglu** [comment]
Thanks for the quick reply. I think setting it to one in these cases is probably the best way to go (but i understand that is problematic for backwards compability). A warning and certainly a comment in the docs would be fine too. Ignore_zeros would be reasonable too. Alternatively, one could say that if variance gets smaller than user defined parameter the variance is set to one or a user defined argument maybe. There is probably even some theory which suggests what to do in these cases. 
 
Regarding the `partial_adapt` . I think implementing this by myself would be straightforward at this point so not sure if its useful fo others.

I  spent quite a good time on this standarzation issue and my key takeaways are:

- Doing this in Pytorch is also extremely cumbersome: [Link](https://discuss.pytorch.org/t/extremely-ugly-but-only-way-to-standardize-dataset/206301/2)

- Partial_adapt would be reasonable for very large datasets but one could also just precompute the mean and var once and store it for the dataset in question (in a txt file). For deep learning partial_adapt is sensible (after one epoch we would have the correct mean and var and simplifies the code above alot) but for other methods it may be crucial to get it right on the "first pass". 

- Finally one could put the mean and var in the meta/ info section of the datasets.

So i think updating the docs would be the most important thing. For the other options you may have to add a bunch of stuff so not sure if it makes sense.

**Comment by grasskin** [comment]
After some internal discussion we'll be leaving this as is for compatibility but welcome any PR's adding documentation (and of course this issue will be searchable)!

On `partial_adapt` I agree the use case isn't as clear since larger datasets you would likely want to precompute mean and variance anyway and for smaller datasets this isn't an issue. That being said follow-up PR's welcome.

Thank you for opening this @yCobanoglu, closing for now but feel free to reopen if we implement `partial_adapt`/for follow-ups.
